### PR TITLE
GDScript enum fixes & refactor

### LINKED ISSUE #57966: GDScript 2.0: Function return type `enum` error (Trying to return value of type "int" from a function which the return type is "Dictionary")
### Godot version

v4.0.alpha2.official [79077e6c1]

### System information

5.13.19-2-MANJARO XFCE, Nvidia 2060

### Issue description

Work fine in alpha 1 (v4.0.alpha1.official  @[31a7ddbf8])

Error exists in alpha 2 (v4.0.alpha2.official [79077e6c1])

Also Auto-Completion does not show the error.

### Steps to reproduce

```
enum State {
	Idle,
	Run,
	Attack,
}


func _ready() -> void:
	getEnum()

func getEnum() -> State:
	return State.Run
```

error: `Trying to return value of type "int" from a function which the return type is "Dictionary".`

### Minimal reproduction project

_No response_

### LINKED ISSUE #60545: GDScript 2.0 : When using Named Enums, accessing invalid members won't be caught while editing
### Godot version

v4.0.alpha6.official [e4f0fc50f]

### System information

Windows 10

### Issue description

If you access an invalid member of a named enum, the editor won't catch the issue, but an error will be thrown at runtime.
If you do the same thing with an unnamed enum, the editor will flag it immediately.

I understand this is likely because named enums are a Dictionary, but it seems to me that it would be an improvement having named enums get flagged in the editor similar to unnamed enums.

### Steps to reproduce

See code in repro section

### Minimal reproduction project

enum GAME_STATE {WIN, LOSE}

var state1_good = GAME_STATE.WIN
var state1_bad = GAME_STATE.DRAW  #This will throw an error at runtime

enum {STATE_WIN, STATE_LOSE}

var state2_good = STATE_WIN
var state2_bad = STATE_DRAW #This will be flagged in the editor

### LINKED ISSUE #61427: Named enums does not provide hints
### Godot version

v4.0.alpha8.official [cc3ed63af]

### System information

Windows 10

### Issue description

Like in the title. Every named enum that I create in 4.0 lacks of any hints for typing. All it hints is using 'new' method on it (which causes several errors while executing)
Enum1
Enum2
.

### Steps to reproduce

Create named enum and try to assign a value from it.

### Minimal reproduction project

_No response_

### LINKED ISSUE #61466: Calling enum's nonexisting method causes a crash
### Godot version

v4.0.alpha8.official [cc3ed63af]

### System information

Windows 10, OpenGL ES 3.0 Renderer: GeForce GTX 1060 6GB/PCIe/SSE2

### Issue description

Trying to use enum name as something that has methods, crashes Godot:

kuva

### Steps to reproduce

```gdscript
extends Node2D

enum Foo {abc, def}

func _ready():
	Foo.garbage()
```

### Minimal reproduction project

_No response_

### LINKED ISSUE #69118: Enum initialized export generates incorrect warning
### Godot version

4.0 beta 6

### System information

Linux

### Issue description

When exporting a variable of an enum type, in this case: 

```GDScript
@export var tween_ease: Tween.EaseType = Tween.EASE_OUT
```
The console generates the following warning:

```
W 0:00:00:0329   Integer used when an enum value is expected. If this is intended cast the integer to the enum type.
  <Error de C++> INT_ASSIGNED_TO_ENUM
  <Fuente>       custom_progress_bar.gd:6
```

Which doesn't really make any sense as no Integer is being cast (?)

### Steps to reproduce

1. Create a GDScript script
2. Declare and initialize an enum variable as specified above
3. Run the code and check the warnings

### Minimal reproduction project

N/A

### LINKED ISSUE #70915: Stack overflow when using @export variable with enum type and set, get
### Godot version

4.x alpha2 - now (dc3175e)
Last worked at alpha1.
```
Alpha1
<--- Regression introduced here
Alpha2
```

### System information

Windows 10

### Issue description

When declaring an `@export` variable with an enum type and `set, get`, Godot will first throw an error and then result in a stack overflow.

The initial error message is: `Trying to return value of type "int" from a function which the return type is "Dictionary"`
Then the stack overflow:
image

When removing the `@export` annotation, no stack overflow happens anymore but the initial error is still there and the program will not launch.

### Steps to reproduce

1. Either open or copy attached project/code.
2. Launch
3. See the error and after a while the stack overflow

### Minimal reproduction project

[Godot4Bug.zip](https://github.com/godotengine/godot/files/10344392/Godot4Bug.zip)

---

class `B`:
```
class_name B
extends Node

enum Faction {
	ONE, TWO, THREE
}

@export var faction: Faction = Faction.ONE : set = set_faction, get = get_faction

func _ready():
	set_faction(Faction.TWO)
	pass

func set_faction(new_faction: Faction):
	print(faction)
	faction = new_faction

func get_faction() -> Faction:
	return faction
```
main script:
```
var b: B = B.new() 
add_child(b)
```

### PR DESCRIPTION
**Pull request by anvilfolk**
This PR fixes a bunch of issues with enums. Also, 80% of it is unit tests ;)

~~Needs #71028 before it passes tests :)~~

### At a glance

- Fix being able to assign to enum values.
- Fix unsafe enum assignments not marking nodes unsafe.
- Fix native enum exports generating warning. Fixes #69118.
- Fix enum casts with values not existing in enum target not yielding error.
- Fix no errors being generated when using non-existent enum values. Fixes #60545.
- Fix accessing non-existent properties of native types (including enums) not yielding error.
- Fix no errors being generated when using non-existent methods on enums. Fixes #61466.
- Fix being able to use non-const dictionary methods on enum types, e.g. `MyEnum.clear()`.
- Fix code autocompletion for enums. Only enum values and const dictionary methods show up now. Fixes #61427.
- Fix (native & gdscript) enum values not being accessible in certain ways, e.g. `NativeClass.EnumType.ENUM_VAL`, etc.
- Fix enum and native enum types not being usable as variable, function parameter, function return types. Fixes #57966.
- Fix #70915.

- Added comprehensive unit tests.
- Improved error messages with more information.
- Code refactor & other assorted improvements and fixes.

And yes, maybe I _did_ name a lot of these "fixes" instead of "implements" because we are in feature freeze ;)

### Fixes & features

It does a general refactor of how enums are handled internally and in GDScript, cleaning up the code & making it slightly more consistent across GDScript enums and native enums. It also aims to unify warnings&errors between in-editor, runtime, and export debug/release templates (native enum behavior on release templates now working due to #64253).

For native enums, e.g. with `TileSet`, you can now use `TileSet.TILE_SHAPE_SQUARE` or `TileSet.TileShape.TILE_SHAPE_SQUARE`, and a warning will be issued if the native type does not have an accessed property.

Typing with enums now works properly for variables, function parameters and function returns, and comprehensive unit tests have been added to check this. Example which is now possible, and will throw errors if the wrong enum is used.

```GDScript
class_name OuterClass

enum MyEnum {V0, V1}

class InnerClass:
  enum MyEnum {V0, V1}

func outer_inner_class_class(e : InnerClass.MyEnum) -> Outer.InnerClass.MyEnum:
  print(e)
  return e

func _ready()
  var boom :  = outer_inner_class_class(V1)
  var good :  = outer_inner_class_class(OuterClass.InnerClass.V1)
```

### CODE REVIEW

**Review by rune-scape — COMMENTED** [review]
i know it's still a draft, but this looks great! nice job refactoring :)

**Comment by rune-scape** [comment]
what happens with this script?:
```gdscript
enum E1 {V0, V1}
enum E2 {V1, V0}

func _ready():
    print(V0, V1)
```

**Comment by anvilfolk** [comment]
Oooooh, excellent comments, thanks @rune-scape! I am done for the day, but will think & tackle these when I can! I've implemented everything I wanted for this PR, so your comments are super timely :)

File: modules/gdscript/gdscript_analyzer.cpp [hunk 1]
```diff
@@ -106,13 +106,26 @@ static GDScriptParser::DataType make_native_meta_type(const StringName &p_class_
...
 	type.kind = GDScriptParser::DataType::ENUM;
-	type.builtin_type = Variant::INT;
+	type.builtin_type = p_meta ? Variant::DICTIONARY : Variant::INT;
+	type.enum_type = p_enum_name;
 	type.is_constant = true;
-	type.is_meta_type = true;
+	type.is_meta_type = p_meta;
+
+	if (!p_base_name.is_empty()) {
+		type.native_type = p_base_name + "::" + p_enum_name;
```
[the remark this answers is not in the document]
Comment (anvilfolk) [reply modules/gdscript/gdscript_analyzer.cpp:121]: This was chosen mostly for aesthetic reasons!

Some classes are defined by their filename, so when printing errors, you would end up with awkward enum names like `c:/my_local_hard_drive/test_project/my_class_gdscript_file.gd.MyEnum`. So instead, by using `::`, when printing errors you throw away the path, keep the file, and end up with `my_class_gdscript_file.gd::MyEnum`, which I thought was a lot more readable.

Technically speaking, for enums, `native_type` is only used to compare `DataTypes` and check if enum meta_types match. As long as they uniquely identify their enum, there shouldn't be any issues I believe!

File: modules/gdscript/gdscript_analyzer.cpp [hunk 2]
```diff
@@ -2772,14 +2827,9 @@ void GDScriptAnalyzer::reduce_identifier_from_base(GDScriptParser::IdentifierNod
...
-				GDScriptParser::DataType result;
-				result.type_source = GDScriptParser::DataType::ANNOTATED_EXPLICIT;
-				result.kind = GDScriptParser::DataType::ENUM;
-				result.is_constant = true;
+				GDScriptParser::DataType result = base;
 				result.builtin_type = Variant::INT;
-				result.native_type = base.native_type;
-				result.enum_type = base.enum_type;
-				result.enum_values = base.enum_values;
+				result.is_meta_type = false;
```
[the remark this answers is not in the document]
Comment (anvilfolk) [reply modules/gdscript/gdscript_analyzer.cpp:2832]: Yessss, great call! Will update!

File: modules/gdscript/gdscript_analyzer.cpp [hunk 3]
```diff
@@ -2647,11 +2676,37 @@ void GDScriptAnalyzer::reduce_cast(GDScriptParser::CastNode *p_cast) {
...
 			} else if (op_type.kind == GDScriptParser::DataType::BUILTIN && op_type.builtin_type == Variant::INT && cast_type.kind == GDScriptParser::DataType::ENUM) {
-				// Convertint int to enum is always valid.
-				valid = true;
+				// Int assignment to enum not valid when exact int assigned is known but is not an enum value
+				if (p_cast->operand->is_constant && p_cast->operand->reduced) {
+					if (enum_get_value_name(cast_type, p_cast->operand->reduced_value) != StringName()) {
+						valid = true;
+					} else {
+						valid = false;
+						push_error(vformat(R"(Invalid cast. Enum "%s" does not have enum number %d.)", cast_type.to_string(), p_cast->operand->reduced_value.operator uint64_t()), p_cast->cast_type);
```
[the remark this answers is not in the document]
Comment (anvilfolk) [reply modules/gdscript/gdscript_analyzer.cpp:2703]: Oh, good call. I'll look into it! :)

**Comment by anvilfolk** [comment]
> what happens with this script?:

That's a really cool one!

That's where trying to keep things consistent & backwards compatible came in. Firstly, for native classes, `NativeClass.EnumName.ENUM_VALUE` wasn't possible, only `NativeClass.ENUM_VALUE`. On the other hand, GDScript was the opposite, with `GDScriptClass.NAMED_ENUM_VALUE` not working, only `NamedEnum.NAMED_ENUM_VALUE`.

So the two behaviors existed for two classes & and made it super inconsistent. I figured the only way to move forward without breaking existing code was to allow both types of access to both types of enums.

Good thought on maybe using `Member` to represent named enum values. I thought about it briefly, but I believe we would lose type information if we created `ENUM_VALUE` members from `ENUM` values. I'm tempted to just add checking named enums wherever appropriate once I factor that out into a function.

But now you're right, it creates issues with shadowing that need some work. GDScript seems pretty lenient on allowing shadowing with warnings rather than errors, so I'm tempted to implement this:

1. If an unnamed enum value shadow named enum values, since named enums can be accessed by adding EnumName qualifiers.
2. Local enum values, unnamed & named, will shadow `base_class` & `outer_class` enum values of any kind, since "upper" enums can be accessed by adding class qualifiers.

I'm tempted to either not add any warnings at all, or to add warnings to 2 only, since it might be less obvious that you are doing it.

File: modules/gdscript/gdscript_analyzer.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (adamscott) [reply modules/gdscript/gdscript_analyzer.cpp:121]: > so when printing errors, you would end up with awkward enum names like `c:/my_local_hard_drive/test_project/my_class_gdscript_file.gd.MyEnum`.

That's only the case for the tests. Otherwise, there will mostly always have the `res://` prefix to the GDScript files.

[the remark this answers is not in the document]
Comment (anvilfolk) [reply modules/gdscript/gdscript_analyzer.cpp:121]: @adamscott yes!!! Except that when you run your tests from the executable, it uses the entire path! So if my path is different than wherever GitHub runs Godot tests, it will fail. Probably to do with how the unit test framework loads GDScripts?

**Comment by anvilfolk** [comment]
All that's left is adding shadow warnings :)

File: modules/gdscript/gdscript_analyzer.cpp [hunk 4]
```diff
@@ -2861,38 +2927,55 @@ void GDScriptAnalyzer::reduce_identifier_from_base(GDScriptParser::IdentifierNod
 					p_identifier->set_datatype(member.constant->initializer->get_datatype());
 					p_identifier->source = GDScriptParser::IdentifierNode::MEMBER_CONSTANT;
 					p_identifier->constant_source = member.constant;
-					break;
+					return;
```
Comment (anvilfolk) [inline modules/gdscript/gdscript_analyzer.cpp:2930]: I replaced a bunch of breaks here by returns - all the unit tests pass, but could someone else double-check that this is safe? I feel like it should be. I don't think any extra work is done after this switch.

**Comment by anvilfolk** [comment]
I actually realized that I should be able to use `reduce_identifier_from_base` to find named enum values, rather than using my auxiliary methods. 

So that + adding shadow warnings are on my list, but hopefully that's it?

File: modules/gdscript/gdscript_analyzer.cpp (hunk 4 above)
Comment (rune-scape) [reply modules/gdscript/gdscript_analyzer.cpp:2930]: i don't see why? theres a return statement on line 2956
the previous code flow looks better to me because its clear that if the class has the member, it always returns

Comment (anvilfolk) [reply modules/gdscript/gdscript_analyzer.cpp:2930]: You are totally right! Clearly haven't been sleeping well and absolutely missed it. Thank you!

**Comment by anvilfolk** [comment]
Alright, I think this PR needs to be done :)

There's some issues with shadowing outer class members that needs work in `check_class_member_name_conflict` but I think that's a different PR.

File: modules/gdscript/gdscript_analyzer.cpp (hunk 4 above)
Comment (rune-scape) [reply modules/gdscript/gdscript_analyzer.cpp]: sleep is the most important thing ❤️

**Comment by anvilfolk** [comment]
Will work on rebasing this one :)

**Comment by anvilfolk** [comment]
Should be ready to merge :)

**Comment by rune-scape** [comment]
i'm still not for the change to make named enums values accessible in the class scope, i don't see a reason for it to change from 3.x (and personally i wouldn't expect it to work like that)
with this script and this PR:
```gdscript
const VALUE = 42

class Inner:
    enum MemberEnum1 { VALUE = 1 }
    enum MemberEnum2 { VALUE = 2 }
    enum MemberEnum3 { Node = 3 }

    class Inner2:
        const V1 = VALUE
        const V2 = Node
```
i would say the MemberEnum1 value clashes with the MemberEnum2 value not shadows it, since they're declared in the same scope
i think `V1` would be set to `MemberEnum1.VALUE` here?
and would `V2` be set to `MemberEnum3.Node`?
im also pretty sure the compiler or the GDScript class itself support getting enum values like this, so this would just be a compile time thing

**Comment by anvilfolk** [comment]
Oh, and to actual answer your questions & clarify how I personally think things should work! :)

- `enum MemberEnum1 { VALUE = 1 }` should throw a clash error since now you can't reach outer class `VALUE`.
- `enum MemberEnum2 { VALUE = 2 }` should at least throw a shadow warning for `MemberEnum1.VALUE`. I think a warning is enough since disambiguation can always be done with `MemberEnum1.VALUE`.
- `enum MemberEnum3 { Node = 3 }` should throw a clash error since `Node` is a builtin type.

This effectively makes named enums behave the same as regular enums. We just need to make sure the warnings & errors are there.

But before putting in some of these, I'd like to see the results of #70246 so the warnings are consistent with that behavior too :)

**Comment by anvilfolk** [comment]
Since there is some pushback, here is the proposal to discuss it :)

https://github.com/godotengine/godot-proposals/issues/5968

**Comment by anvilfolk** [comment]
There is clearly not a lot of support named enum values being scoped to their class, so I am setting this as draft while I remove that feature, add new unit tests to make sure that behavior is well documented.

I'm also going to have to fix some issues where the `GDScriptAnalyzer` behaves differently when generating unit test outputs vs in-editor/running a scene. Very strange :)

**Comment by anvilfolk** [comment]
Should once again be ready to go, though there seem to be some regressions needing more urgent attention at the moment :)

File: modules/gdscript/gdscript_parser.cpp [hunk 5]
```diff
@@ -4087,8 +4089,15 @@ String GDScriptParser::DataType::to_string() const {
 			}
 			return native_type.operator String();
 		}
-		case ENUM:
-			return enum_type.operator String() + " (enum)";
+		case ENUM: {
+			if (native_type != StringName()) {
+				// If native type is a file, hide file path and keep only file name
```
Comment (adamscott) [inline modules/gdscript/gdscript_parser.cpp:4094]: I'm wondering. If my file name is `A.gd` and I import it by doing `const B = preload("A.gd")`, will it write `A` as the file type?

**Review by adamscott — APPROVED** [review]
Looks good to me! There's a lot of tests, I'm confident. Good job, @anvilfolk, you really did well.

File: modules/gdscript/gdscript_parser.cpp (hunk 5 above)
Comment (anvilfolk) [reply modules/gdscript/gdscript_parser.cpp:4094]: This is just for enums, so I'm not sure! So you mean preloading an external class that has an enum? Probably should add a test for that.

Something like:

```GDScript
const B = preload("A.gd")
make_error_with(B.SomeEnum)
```

Comment (anvilfolk) [reply modules/gdscript/gdscript_parser.cpp:4094]: Adding the following test:

```GDScript
enum LocalNamed { VALUE_A, VALUE_B, VALUE_C = 42 }

func test():
	const P = preload("../features/enum_from_outer.gd")
	var x : LocalNamed
	x = P.Named.VALUE_A
```

Output:

```GDTEST_ANALYZER_ERROR
Value of type "enum_from_outer.gd::Named" cannot be assigned to a variable of type "preload_enum_error.gd::LocalNamed".
```

**Comment by anvilfolk** [comment]
Some changes I made appear to have caused issues with cyclic dependency code #67714, which is why the tests are failing. It appears to be trying to free some things twice.

Having a really hard time seeing how this code could do that, but something's messing up for sure :) Hopefully can collab with @adamscott to figure it out.

File: modules/gdscript/gdscript_parser.cpp (hunk 5 above)
Comment (adamscott) [reply modules/gdscript/gdscript_parser.cpp:4094]: I mean this:
```gdscript
# b.gd
enum {
    VALUE_A,
    VALUE_B,
    VALUE_C
}
```
```gdscript
# test.gd
enum LocalNamed {
    VALUE_A,
    VALUE_B,
    VALUE_C
}

func test():
    const A = preload("b.gd")
    var x: LocalNamed
    x = A.VALUE_A
```

Comment (anvilfolk) [reply modules/gdscript/gdscript_parser.cpp:4094]: oh, interesting, I'm curious to see the error message even without the `preload`! Not sure how named & unnamed enums interact.

Comment (anvilfolk) [reply modules/gdscript/gdscript_parser.cpp:4094]: I'd forgotten that I had accounted for this :)

```GDTEST_ANALYZER_ERROR
Value of type "enum_value_from_parent.gd::UNNAMED_ENUM" cannot be assigned to a variable of type "enum_preload_unnamed_assign_to_named.gd::MyEnum".
```

Added tests for this :)

**Comment by kuhnchris** [comment]
> 

Oh, I compiled your fork yesterday and was wondering why it was crashing sometimes - I'll try to isolate the issue and create a minimal reproduction project and put it in here if you want to take a look at it.
Also, thanks for working on this, it's really great having auto-complete/suggestions for enums.
image

**Comment by adamscott** [comment]
@kuhnchris By checking your image, the suggestions shouldn't include the dictionary methods.

File: modules/gdscript/gdscript_parser.cpp (hunk 5 above)
Comment (adamscott) [reply modules/gdscript/gdscript_parser.cpp:4094]: I wonder if it would be better to name `UNNAMED_ENUM` something like `<enum>`, so it reads: `b.gd::<enum>`.

I have the impression that otherwise, it could conflict with
```gdscript
# enum_value_from_parent.gd
enum UNNAMED_ENUM {
    IM_SO_EDGY,
    WITH_MY_IRONIC_NAME
}
```

Comment (adamscott) [reply modules/gdscript/gdscript_parser.cpp:4094]: Also, I think it should use the entire resource path for the reporting, as there could be multiple `enum_value_from_parent.gd` files in a project.

Comment (anvilfolk) [reply modules/gdscript/gdscript_parser.cpp:4094]: I'll change it to something like `<unnamed enum>` - I would like to make it explicit that it's an enum without names, not just any enum, but you're right, the `<>` is a good indicator!

The issue with having the path is that the path is absolute, so if I generate tests locally, they won't pass on github. Perhaps I could find a way to change the path to a relative path though. 

But people would need to define two .gd files with the same name but different paths, both with unnamed enums. And then get their enums confused? I feel like that's enough of an edge case that maybe it's not worth the longer/more confusing error message in most cases we expect. What do you think?

Comment (adamscott) [reply modules/gdscript/gdscript_parser.cpp:4094]: > The issue with having the path is that the path is absolute, so if I generate tests locally, they won't pass on github. Perhaps I could find a way to change the path to a relative path though.

Why not the `res://` path?

Comment (anvilfolk) [reply modules/gdscript/gdscript_parser.cpp]: The datatype are constructed using the class' fully qualified class name. I'm not sure we do have access to the `res://` path in the analyzer, though it'd be neat if we did?

**Comment by anvilfolk** [comment]
@kuhnchris I didn't really touch the code completion, but I'm glad that stuff now shows up if it didn't before, hehe! I think I'd like to fix what @adamscott says in a subsequent PR, not this one necessarily. I just want to see this get in without any more change of introducing regressions :)

**Comment by kuhnchris** [comment]
Well, since you made them "accessable", they show up in the LSP, I assumed. ;-)
We do have troubles with Unnamed Enums, but I... don't really know what I was expecting, trying to assign a variable of the unnamed enum type fails, but I think that ties into the discussion you guys had above.
I wasn't able to reproduce the crash yet, so I cannot throw up a example project, but did some tests (I assume those are similar to the ones in the tests anyways). - anyways, great work you guys, if there's anything we can help with, let us know 👍  :-)
[Video.webm](https://user-images.githubusercontent.com/761911/211162376-2781a576-593b-4937-8f43-81d25de3d3e4.webm)
(Tests made before commit 425b653)

**Comment by anvilfolk** [comment]
@kuhnchris thanks for testing, it's super helpful!!!

So the main issue for anonymous/unnamed enums is basically that you cannot really typecheck them. Since you don't have an enum name, you can only identify the class that it belongs to, but then that type is a class, not an enum. The errors you are getting, as far as I can see, are telling you exactly that: you're trying to assign an enum value to a variable that's typed as a class.

```GDScript
class EnumHolder:
  enum { V1, V2, V3}
```

You can't really do `var v : EnumHolder = V1` because then `v` expects something like `EnumHolder.new()`, whereas `V1` is basically an int :)

It was a little hard to follow everything you did in the video - but does this make sense? Or did I miss something that is not working as expected? I mostly focused on named enums for this PR, so I'm sure there's going to be something with unnamed enums I've missed :)

**Comment by anvilfolk** [comment]
Also, with #71028 I get the following.

image

So once that gets merged, this should be ready to go! Can't for the life of me figure out what in my code would trigger those errors, but I'm glad they get fixed :)

File: modules/gdscript/gdscript_analyzer.cpp [hunk 6]
```diff
@@ -106,13 +109,26 @@ static GDScriptParser::DataType make_native_meta_type(const StringName &p_class_
...
-	type.builtin_type = Variant::INT;
+	type.builtin_type = p_meta ? Variant::DICTIONARY : Variant::INT;
+	type.enum_type = p_enum_name;
 	type.is_constant = true;
-	type.is_meta_type = true;
+	type.is_meta_type = p_meta;
+
+	// For enums, native_type is only used to check compatibility in is_type_compatible()
+	// We can set anything readable here for error messages, as long as it uniquely identifies the type of the enum
+	type.native_type = p_base_name + ENUM_SEPARATOR + p_enum_name;
```
[the remark this answers is not in the document]
Comment (anvilfolk) [reply modules/gdscript/gdscript_analyzer.cpp:125]: Good call, hopefully that won't change in the future. I'll add a comment just in case :)

File: modules/gdscript/gdscript_analyzer.cpp [hunk 7]
```diff
@@ -608,14 +643,18 @@ GDScriptParser::DataType GDScriptAnalyzer::resolve_datatype(GDScriptParser::Type
...
 						result = member.get_datatype();
+						found = true;
 						break;
 					case GDScriptParser::ClassNode::Member::ENUM:
 						result = member.get_datatype();
+						found = true;
 						break;
 					case GDScriptParser::ClassNode::Member::CONSTANT:
 						if (member.get_datatype().is_meta_type) {
+							// TODO can this ever happen? If it can, should we use type_from_metatype()?
```
[the remark this answers is not in the document]
Comment (anvilfolk) [reply modules/gdscript/gdscript_analyzer.cpp:654]: Makes sense now :)

File: modules/gdscript/gdscript_analyzer.cpp [hunk 8]
```diff
@@ -1330,7 +1361,7 @@ void GDScriptAnalyzer::resolve_function_signature(GDScriptParser::FunctionNode *
 		}
 	} else {
 		if (p_function->return_type != nullptr) {
-			p_function->set_datatype(resolve_datatype(p_function->return_type));
+			p_function->set_datatype(type_from_metatype(resolve_datatype(p_function->return_type)));
```
[the remark this answers is not in the document]
Comment (anvilfolk) [reply modules/gdscript/gdscript_analyzer.cpp:1364]: Agreed! :)

File: modules/gdscript/gdscript_parser.cpp [hunk 9]
```diff
@@ -4087,8 +4089,13 @@ String GDScriptParser::DataType::to_string() const {
...
 			return native_type.operator String();
 		}
-		case ENUM:
-			return enum_type.operator String() + " (enum)";
+		case ENUM: {
+			// native_type contains either the native class defining the enum
+			// or the fully qualified class name of the script defining the enum
+			String native_str = native_type.operator String();
+			String file = native_str.get_file();
+			return (file == native_str) ? native_str : file;
```
[the remark this answers is not in the document]
Comment (anvilfolk) [reply modules/gdscript/gdscript_parser.cpp:4097]: If I remember correctly, it depends on whether `class_name` was defined, at which point it uses that. Otherwise it's the entire filepath and you need to isolate the filename only!

[the remark this answers is not in the document]
Comment (vonagam) [reply modules/gdscript/gdscript_parser.cpp:4097]: I mean, `(x == y) ? y : x` result will always be value of `x`. No?

[the remark this answers is not in the document]
Comment (anvilfolk) [reply modules/gdscript/gdscript_parser.cpp]: Yeah, passes all tests :) Thanks for noticing that!

**Comment by Maran23** [comment]
This may fixes https://github.com/godotengine/godot/issues/70915 as well. Can you verify that? :)

**Comment by anvilfolk** [comment]
It actually does - I confirmed the bug is still there on Beta 10, but doesn't happen in this PR :) Thanks for linking!

**Comment by Maran23** [comment]
> It actually does - I confirmed the bug is still there on Beta 10, but doesn't happen in this PR :) Thanks for linking!

Glad to her!
Not sure if it works, but is it possible to add the buggy example in the ticket as test here?

**Comment by anvilfolk** [comment]
> Not sure if it works, but is it possible to add the buggy example in the ticket as test here?

Not with the current test framework as far as I can tell. Currently it just attempts to parse scripts. The most you can do is preload, but I don't think there's a scene tree we can use. Maybe I'm wrong though :)

File: modules/gdscript/gdscript_compiler.cpp [hunk 10]
```diff
@@ -213,7 +213,7 @@ static bool _have_exact_arguments(const MethodBind *p_method, const Vector<GDScr
 }
 
 GDScriptCodeGenerator::Address GDScriptCompiler::_parse_expression(CodeGen &codegen, Error &r_error, const GDScriptParser::ExpressionNode *p_expression, bool p_root, bool p_initializer, const GDScriptCodeGenerator::Address &p_index_addr) {
-	if (p_expression->is_constant) {
+	if (p_expression->is_constant && !p_expression->get_datatype().is_meta_type) {
```
Comment (anvilfolk) [inline modules/gdscript/gdscript_compiler.cpp:216]: Actually, I just noticed that this is still in here and I'm getting this error in tests if it's not. It was part of why I created #70569, which was supposedly superseded by something else that didn't fix this particular issue (but fixed another).

Now I'm second guessing this, because it means it's not using the reduced_value dictionary.

```
modules\gdscript\tests\gdscript_test_runner.cpp(209): ERROR: CHECK( result.passed ) is NOT correct!
  values: CHECK( false )
  logged: C:/dev/godot/modules/gdscript/tests/scripts/analyzer/features/enum_function_typecheck.gd
          GDTEST_RUNTIME_ERROR
>> SCRIPT ERROR
>> on function: EnumFunctionTypecheckOuterClass.test()
>> analyzer/features/enum_function_typecheck.gd
>> 98
>> Invalid call. Nonexistent function 'inner_inner_no_class' in base 'RefCounted (InnerClass)'.
```

Comment (vonagam) [reply modules/gdscript/gdscript_compiler.cpp:216]: I have not run your branch. But I would say it is using enum's reduced_value, just has to do a lookup instead of getting it immediately.

The issue, my bet, is not with enums, it is with classes. You have there a combo of a global class plus a nested class. This might be the thing with getting full script for the class.

Comment (anvilfolk) [reply modules/gdscript/gdscript_compiler.cpp:216]: I was mulling it over while running chores and reached the same conclusion. I feel like it's ok to leave that in for now and look for a fuller solution later!

File: modules/gdscript/gdscript_analyzer.h [hunk 11]
```diff
@@ -107,7 +107,7 @@ class GDScriptAnalyzer {
 
 	// Helpers.
 	GDScriptParser::DataType type_from_variant(const Variant &p_value, const GDScriptParser::Node *p_source);
-	GDScriptParser::DataType type_from_metatype(const GDScriptParser::DataType &p_meta_type) const;
+	static GDScriptParser::DataType type_from_metatype(const GDScriptParser::DataType &p_meta_type);
```
Comment (vnen) [inline modules/gdscript/gdscript_analyzer.h:110]: Changing this to `static` is not really necessary (though not harmful either). It would a better improvement to actually remove it from the class and keep it as static standalone function in the `cpp` file.

**Review by vnen — APPROVED** [review]
Looks okay to me.

File: modules/gdscript/gdscript_analyzer.h (hunk 11 above)
Comment (anvilfolk) [reply modules/gdscript/gdscript_analyzer.h:110]: Yes, I think at some point a little refactor making sure anything not using Parser or Analyzer state is made class static or file static depending on whether it's used outside a single file makes a lot of sense! Thanks for looking into the PR :)

**Comment by akien-mga** [comment]
Thanks! Outstanding work :tada:
